copr_builder/copr_builder.py

```diff
@@ -0,0 +1,157 @@
+import datetime
+import logging
+import os
+import time
+
+import configparser
+import copr
+
+from copr.client_v2.common import BuildStateValues
+from copr.client_v2.net_client import RequestError
+
+from . import COPR_USER_CONF, COPR_REPO_CONF
+from .errors import CoprBuilderError, CoprBuilderAlreadyFailed
+from .copr_project import CoprProject
+
+
+BUILD_URL_TEMPLATE = "https://copr.fedorainfracloud.org/coprs/%s/%s/build/%s"
+COPR_CONFIG = os.path.expanduser('~/.config/copr')
+
+
+log = logging.getLogger("copr.builder")
+
+
+class CoprBuilder(object):
+
+    def __init__(self, conf_file):
+
+        self.config = configparser.ConfigParser()
+        self.config.read(conf_file)
+
+        self._check_copr_token()
+        self.copr = copr.create_client2_from_file_config()
+
+    def _check_copr_token(self):
+        if not os.path.isfile(COPR_CONFIG):
+            raise CoprBuilderError('Copr configuration file not found.')
+
+        expiration = None
+        with open(COPR_CONFIG, 'r') as f:
+            for line in f:
+                if line.startswith('# expiration date:'):
+                    try:
+                        expiration = datetime.datetime.strptime(line, "# expiration date: %Y-%m-%d\n")
+                    except ValueError:
+                        # parsing failed, just ignore it
+                        pass
+
+        if expiration is None:
+            log.warning('Failed to get Copr token expiration date from config file.')
+            return
+
+        if expiration < datetime.datetime.now():
+            raise CoprBuilderError('Your Copr token has expired. Expiration date: %s. '
+                                   'Please obtain new at https://copr.fedorainfracloud.org/api/ '
+                                   'and save it to %s.' % (expiration.strftime("%Y-%m-%d"), COPR_CONFIG))
+
+    def _check_projects_input(self, projects):
+        wrong = [p for p in projects if p not in self.config.sections()]
+        if wrong:
+            raise CoprBuilderError('Requested project(s) %s not found in config.' % wrong)
```

Comment:
What about 

```
if any(...):
   raise...
```

Reply:
I need list of "wrong" projects for the exception message.